extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,654 @@
+ *  player: !Element
+ *  chip: !Element,
+ *  width: number,
+ *  height: number,
+ *  deviceHeight: ?number,
+ *  custom: ?boolean,
+ * }}
+ */
+export let DeviceInfo;
+
+export class AmpStoryDevToolsTabPreview extends AMP.BaseElement {
+  /** @param {!Element} element */
+  constructor(element) {
+    super(element);
+
+    /** @private  {?string} */
+    this.storyUrl_ = null;
+
+    /** @private {!Array<DeviceInfo>} */
+    this.devices_ = [];
+  }
+
+  /** @override */
+  buildCallback() {
+    const styleTag = this.element.ownerDocument.createElement('style');
+    styleTag.textContent = CSS;
+    this.element.appendChild(styleTag);
+
+    this.storyUrl_ = this.element.getAttribute('story-url');
+    this.element.classList.add('i-amphtml-story-dev-tools-tab');
+    const chipListContainer = this.element.ownerDocument.createElement('div');
+    chipListContainer.classList.add(
+      'i-amphtml-story-dev-tools-device-chips-container'
+    );
+    this.element.appendChild(chipListContainer);
+
+    const chipList = this.element.ownerDocument.createElement('span');
+    chipList.classList.add('i-amphtml-story-dev-tools-device-chips');
+    chipListContainer.appendChild(chipList);
+    chipList.addEventListener('click', (event) => {
+      const chip = closest(event.target, (el) => el.device);
+      if (chip) {
+        this.removeDevice_(chip.device);
+        this.updateDevicesInHash_();
+      }
+    });
+
+    chipListContainer.appendChild(this.buildAddDeviceButton_());
+    chipListContainer.appendChild(this.buildHelpButton_());
+
+    parseDevices(
+      this.element.getAttribute('devices') || DEFAULT_DEVICES
+    ).forEach((device) => {
+      this.addDevice_(device);
+    });
+  }
+
+  /**
+   * Builds the add device button
+   * @private
+   * @return {!Element}
+   */
+  buildAddDeviceButton_() {
+    const addDeviceButton = buildDeviceChipTemplate(this.element);
+    addDeviceButton.classList.add('i-amphtml-story-dev-tools-add-device');
+    addDeviceButton.classList.add('i-amphtml-story-dev-tools-button');
+    addDeviceButton.querySelector('span').textContent = 'ADD DEVICE';
+    addDeviceButton.addEventListener('click', () => {
+      this.showAddDevicePopup_();
+    });
+    return addDeviceButton;
+  }
+
+  /**
+   * Builds the add device button
+   * @private
+   * @return {!Element}
+   */
+  buildHelpButton_() {
+    const addHelpButton = buildHelpButtonTemplate(this.element);
+    addHelpButton.addEventListener('click', () => {
+      this.showHelpPopup_();
+    });
+    return addHelpButton;
+  }
+
+  /**
+   * Creates a device layout for preview
+   * @private
+   * @param {!DeviceInfo} device
+   * @return {!Element}
+   */
+  buildDeviceLayout_(device) {
+    const deviceLayout = buildDeviceTemplate(this.element);
+    const devicePlayer = deviceLayout.querySelector('amp-story-player');
+    devicePlayer.setAttribute('width', device.width);
+    devicePlayer.setAttribute('height', device.height);
```

Comment:
mutateElement

Reply:
I think this buildDeviceLayout doesn't need mutateElement because it's not attached when it's being built, but the `addDevice_` calls the `appendChild` in the mutate context, so it should work well